pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -185,6 +185,44 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 	return builder
 }
 
+func buildNameToServiceMapForHttpRoutes(push *model.PushContext,
+	virtualService config.Config) map[host.Name]*model.Service {
+
+	vs := virtualService.Spec.(*networking.VirtualService)
+	nameToServiceMap := map[host.Name]*model.Service{}
+
+	addService := func(hostname host.Name) {
+		if _, exist := nameToServiceMap[hostname]; exist {
+			return
+		}
+		// First, we obtain the service from the namespace of virtualService
+		if s, exist := push.ServiceIndex.HostnameAndNamespace[hostname][virtualService.Namespace]; exist {
+			nameToServiceMap[hostname] = s
+		} else {
+			// We fallback to get service based on hostname directly.
+			nameToServiceMap[hostname] = push.ServiceIndex.Hostname[hostname]
```

Comment:
This is not quite right, think about this a SE not exported to this gateway.

Reply:
You are right. I have updated. PTAL.

Reply:
Add more comments for explaining why do this.